apps/files/helpers.py

```diff
@@ -335,30 +342,143 @@ def get_url(self, short):
+
+        main = '%s.js' % harness_options['mainPath'].replace('/', '/lib/')
+        new_main = package.get('main', 'index.js')
+        main_package = main[:main.index('/')]
+
+        have_modules = left.get('node_modules', {}).get('directory')
+        if have_modules:
+            # Need to create a 'node_modules' directory in the right
+            # file tree, just to try to ward off any possible
+            # issues. Easiest thing is to steal it from the left
+            # side.
+            right['node_modules'] = dict(left['node_modules'])
+
+        def move(src, dst):
+            if dst not in right:
+                right[dst] = right[src]
+                del right[src]
+
+        package = None
+        prefix = '/'
+        new_prefix = None
+
+        # No iterators here. We're going to be modifying the dict,
+        # so they'll break.
+        for path in right.keys():
+            if path.startswith('resources/'):
+                p = path[len('resources/'):]
+
+                if p == main:
+                    # Always try to move the old main module to the
+                    # location of the new main, regardless of
+                    # package.
+                    move(path, new_main)
+                elif p.startswith('addon-sdk/') or p == 'addon-sdk':
+                    # No longer used. Just drop.
+                    del right[path]
+                elif '/' not in p:
+                    # Package directory
+                    package = p
+                    prefix = '%s/' % p
+                    if package == main_package:
+                        new_prefix = ''
+                    else:
+                        if have_modules:
+                            new_prefix = 'node_modules/%s/' % p
+                        else:
+                            new_prefix = '%s/' % p
+
+                        move(path, new_prefix[:-1])
+                # This should always be true, but just in case.
+                elif p.startswith(prefix):
+                    # File in a new package. Try to find where it
+                    # belongs.
+                    f = p[len(prefix):]
+
+                    targets = [f]
+                    for pfx in 'lib/', 'data/':
+                        if f.startswith(pfx):
+                            targets.append(f[len(pfx):])
+
+                    for target in targets:
+                        t = ''.join((new_prefix, target))
+                        if t in left:
+                            move(path, t)
+                            break
+
+                    # If no match found, just give up and dump it in
+                    # the package directory.
+                    if path in right and path not in left:
+                        # But strip off the leading 'lib/' first.
+                        if f.startswith('lib/'):
+                            f = f[len('lib/'):]
+
+                        move(path, new_prefix + f)
+                else:
```

Comment:
Why have a specific `else` that does nothing then? Either log, raise, assert... or remove it ;)